### PR TITLE
Add generic type support

### PR DESCRIPTION
**Pull request by DevNullx64**
Add support for float and double.
It's volontary locked to float type only because of unknowing side effect in discrete number.

the main programe use MLP<float>

The network can't learn in with Half. Don't know why.

### CODE REVIEW

**Comment by Leonardo16AM** [comment]
I think the example network is not learning ok because of the gradient descent implementation in the example, I'll take a look at it later today.

**Review by Leonardo16AM — APPROVED** [review]

**Comment by Leonardo16AM** [comment]
Aparently the network  is learning, or at least the loss is decreasing: 

> Epoch: 116
> Loss: 677.8246
> Epoch: 117
> Loss: 619.62286
> Epoch: 118
> Loss: 572.9854
> Epoch: 119
> Loss: 539.3139
> Epoch: 120
> Loss: 514.9062
> Epoch: 121
> Loss: 500.12036
> Epoch: 122
> Loss: 494.77658
> Epoch: 123
> Loss: 499.94394
> Final loss: 499.94394

Maybe the example is not the best one. Apart from that i think that we need to build the NN over tensorial operations, at least if we want to optimize it to run in a GPU,  and then, over that write a solid NN Module with good optimization functions such as Adam or SGD.  Also I think it will be good to add more activation functions (at least Sigmoid, Leaky ReLU, Softmax and Fix TanH) I will open a branch to add those activation functions.

**Comment by DevNullx64** [comment]
I very happy to see that in Half it's work. On my test I'd a loss bump within the  first epoch. I'd not go futher.
I'd test 2 or 3 aproach last weeks to support tensor. At en, it should be a good thing if every object can be structure. ILGPU use that. Futhermore, I think that dimension in shape of a tensor should be typed. I think about that because einstein notation looklike an attempt to identify dimensions. Decide the shape an min what dimension of thit shape identifiy what kinf of information (batch dimension ? height ? with ? Depth ? Other ?). That one of thing that let me turn around a litle...
I think I try to fly before learning to walk. That why I reduce my near objective to vectorized unit, prehaps multithreading then ILGPU.

**Comment by DevNullx64** [comment]
But my goal is to have something that can run forward and backward. Bypassing existing TensorFlow or Torch. This two Library are best for prototyping and try new architecture. But In real, It's overload the VRam band with. ILGPU can be used to compile kernel for layer and why not a model as kernel. Yes... that need more work ^^'

**Comment by Leonardo16AM** [comment]
I think we should create a roadmap to work on, set milestones, and define specific objectives for what we want to achieve. I've created a Github Discussion where I think we can debate this, although I haven't worked with Github Discussions before. 

Regarding the use of ILGPU, I need to research the library since I haven't used it. Another option would be to define CUDA Kernels directly, but that might be unnecessary at this stage. Can you send me the examples you've been testing with?

**Comment by Leonardo16AM** [comment]
Ok, I have now read a bit about ILGPU and it looks fantastic. It seems to me that using it is the right option. I would now like to define step by step what we will do, maybe define the class structure, and the functionalities we would like to implement.
